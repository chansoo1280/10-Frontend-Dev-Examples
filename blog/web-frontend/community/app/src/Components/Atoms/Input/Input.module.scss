@import "@Styles/scss/default.scss";
$input:"input";
.input{
    position:relative;
    display:flex;
    height:32px;
    font-size:14px;
    
    @include click-animating();
    &--large{
        &__input{
            padding:12px 8px;
        }
        height:40px;
        font-size:16px;
    }
    &--small{
        &__input{
            padding:1px 8px;
        }
        height:24px;
        font-size:14px;
    }
    &:hover{
        .#{$input}__input{
            @include set-color(rgba($black, .85), transparent,$primary-5);
        }
    }
    &__input{
        padding:5px 12px;
        width:100%;
        height:100%;
        box-sizing:border-box;
        border:1px solid;
        border-radius:$border-radius;
        outline:none;
        font-size:inherit;
        &::placeholder{
            color:rgba($black, .25);
        }
        transition:background-color .2s ease, border-color .2s ease, color .2s ease;
        @include set-color(rgba($black, .85), $white, $gray-5); 
        &:hover, &:focus{
            @include set-color(rgba($black, .85), transparent,$primary-5);
        }
        &:disabled{
            @include disabled-color();
        }
        &:hover, &:focus{
            &:disabled{
                @include disabled-color();
            }
        }
        
    }
}