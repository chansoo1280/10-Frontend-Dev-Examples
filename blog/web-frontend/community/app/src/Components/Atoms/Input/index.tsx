// #region Global Imports
import React, { ChangeEvent, ChangeEventHandler, ReactNode, useRef, useState } from "react"
import classNames from "classnames"
// #endregion Global Imports

// #region Local Imports
import { defaultProps } from "@Components"
import { useClickAnimating } from "@Hooks"
import styles from "./Input.module.scss"
// #endregion Local Imports

interface BaseInputProps extends defaultProps {
    type?: "text" | "email" | "password"
    size?: "small" | "medium" | "large"
    value: string
    onChange: ChangeEventHandler<HTMLInputElement>
    prefix?: ReactNode
    suffix?: ReactNode
}

type InputProps = BaseInputProps & Omit<React.InputHTMLAttributes<any>, "value" | "type" | "onChange" | "size" | "prefix" | "suffix">
const Input = (props: InputProps): JSX.Element => {
    const { value, show, size, className, onChange, disabled, prefix, suffix, placeholder, ...rest } = props
    const inputRef = useRef<HTMLInputElement>(null)
    const [innerValue, setValue] = useState(value)
    const { isClick, execClickAnimation } = useClickAnimating()
    const prefixCls = "input"
    const classes = classNames(
        styles[`${prefixCls}`],
        {
            [styles[`${prefixCls}--hide`]]: show === false,
            [styles[`${prefixCls}--${size}`]]: size,
        },
        className,
    )
    const inputClasses = classNames(styles[`${prefixCls}__input`])
    const handleClick = () => {
        execClickAnimation()
    }
    const handleClickBox = () => {
        execClickAnimation()
        inputRef?.current?.focus()
    }
    const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
        e.preventDefault()
        if (!onChange || disabled) {
            setValue("")
            return
        }
        onChange(e)
        setValue(e.currentTarget.value)
    }
    return (
        <div className={classes} data-click-animating={isClick === true} {...rest} onClick={handleClickBox}>
            {prefix}
            <input ref={inputRef} className={inputClasses} onChange={handleChange} disabled={disabled} value={innerValue} onClick={handleClick} placeholder={placeholder} />
            {suffix}
        </div>
    )
}
export default Input
