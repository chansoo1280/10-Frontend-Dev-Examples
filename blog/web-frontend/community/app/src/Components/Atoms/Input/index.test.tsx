import { fireEvent, render, screen } from "@testing-library/react"
import Input from "./"
describe("Input", () => {
    it("renders input unchanged", () => {
        const handleChange = jest.fn()
        const { container } = render(<Input value="" onChange={handleChange} />)
        expect(container).toMatchSnapshot()
    })
    it("should render a input with the class", () => {
        const handleChange = jest.fn()
        render(<Input value="" onChange={handleChange} data-testid="input" />)

        const input = screen.getByTestId("input")
        expect(input).toBeInTheDocument()
        expect(input).toHaveClass("input")
    })
    it("텍스트 입력 후 값 확인", () => {
        const handleChange = jest.fn()
        render(<Input placeholder="입력해주세요" value="" onChange={handleChange} data-testid="input" />)

        const input = screen.getByPlaceholderText("입력해주세요")
        fireEvent.change(input, { target: { value: "asdfasdf" } })
        expect(input).toHaveValue("asdfasdf")
    })
    it("비활성화 상태에서 텍스트 입력 후 값 확인", () => {
        const handleChange = jest.fn()
        render(<Input placeholder="입력해주세요" value="" onChange={handleChange} data-testid="input" disabled />)

        const input = screen.getByPlaceholderText("입력해주세요")
        fireEvent.change(input, { target: { value: "asdfasdf" } })
        expect(input).toHaveValue("")
    })
})
