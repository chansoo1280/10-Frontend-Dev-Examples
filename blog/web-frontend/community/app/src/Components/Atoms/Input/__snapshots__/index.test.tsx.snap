// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders input unchanged 1`] = `
<div>
  <div
    class="input"
    data-click-animating="false"
  >
    <input
      class="input__input"
      value=""
    />
  </div>
</div>
`;
